winch/environ/src/lib.rs

```diff
@@ -0,0 +1,41 @@
+//! This crate implements Winch's function compilation environment,
+//! which allows Winch's code generation to resolve module and runtime
+//! specific information.  This crate mainly implements the
+//! `winch_codegen::FuncEnv` trait.
+
+use wasmparser::types::Types;
+use wasmtime_environ::{FuncIndex, Module};
+use winch_codegen::{self, Callee, TargetIsa};
+
+/// Function environment containing module and runtime specific
+/// information.
+pub struct FuncEnv<'a> {
+    /// The translated WebAssembly module.
+    pub module: &'a Module,
+    /// Type information about a module, once it has been validated.
+    pub types: &'a Types,
+    /// The current ISA.
+    pub isa: &'a dyn TargetIsa,
+}
+
+impl<'a> winch_codegen::FuncEnv for FuncEnv<'a> {
+    fn callee_from_index(&self, index: u32) -> Callee {
+        let func = self
+            .types
+            .function_at(index)
+            .expect(&format!("function type at index: {}", index));
```

Comment:
A little efficiency thing: this will always invoke `format!` and allocate the string, even in the happy path; I think maybe a better option here is `.unwrap_or_else(|| panic!(...))` or `.ok_or_else(|| format!(...)).unwrap()`?

Reply:
Good catch -- opted to use `unwrap_or_else(|| ...)`